test/unit-test/catch_assert.h

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/*
+ * How to catch an assert:
+ * - save a jump buffer where execution will resume after the assert
+ * - setup a handler for the abort signal, call longjmp within
+ * - optional - close stderr ( fd 2 ) to discard the assert message
+ *
+ * Unity also does a longjmp within its TEST_ASSERT* macros,
+ * so the macro below restores stderr and the prior abort handler
+ * before calling the Unity macro.
+ */
+
+#ifndef CATCH_ASSERT_H_
+#define CATCH_ASSERT_H_
+
+#include <setjmp.h>
+#include <signal.h>
+#include <unistd.h>
+
+#ifndef CATCH_JMPBUF
+    #define CATCH_JMPBUF    waypoint_
+#endif
+
+jmp_buf CATCH_JMPBUF;
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-function"
+static void catchHandler_( int signal )
+{
+    longjmp( CATCH_JMPBUF, signal );
+}
+#pragma GCC diagnostic pop
+
+#define catch_assert( x )                    \
```

Comment:
Why do we need this `setjmp()/longjmp()` magic?  Would it be possible to simplify redefine `assert()` in the unit test to set a variable that the unit test code checks?

Reply:
Thanks for the review.
In some cases, setting a variable is not enough. The code is supposed to abort in such situations. Such as:
```
configASSERT( MyPtr != NULL );
MyPtr->member = 2;
```
In the above case, the `setjmp` magic would work. But if we had just set a variable, then this will cause a segmentation fault (since MyPtr is NULL).

Reply:
makes sense, thank you